### ISSUE TITLE
Implement ok_or() for CtOption<T>

### ISSUE THREAD

**Issue by CPerezz** [dalek-cryptography/subtle#74]
Currently, working with `bls12_381` we have functions like `Scalar::from_bytes()` which returns a `CtOption`. 

This is fine since it's "marking" the function as a const_time fn. The problem is that the type is really unflexible since you don't have any way to transform the `CtOption` to `Result<T, Error>` appart of doing something like:
```rust
if ct_opt.is_some() {
     return ct_opt.unwrap()
} else {
    return Err(....)
}
```

Is there any reason why `ok_or()` it's not implemented for `CtOption`? Or any workarround to that?

**Comment by hdevalence** [comment]
Yeah, this part of the API kinda sucks right now.  I think that the cleanest way to fix this is to do #73 and provide an `Option` conversion -- then your example could be written as
```rust
Option::from(ct_opt).ok_or(...)
```

Implementing `From` gives an automatic `Into` impl, but it might not always work to call `into()` because it may not be possible to infer the return type.  So there might be an ergonomic benefit to also providing `CtOption<T>::into_option(self) -> Option<T> { self.into() }`.  Then I think it would be possible to write

```rust
let s = Scalar::from_bytes(&bytes).into_option().ok_or_else(|| MyError)?;
```

I'm not totally sure on what the naming of `into_option` should be -- it would be good to be consistent with the conversion from `Choice` to `bool`.

I'm really busy right now so I can't do this myself but I'd be happy to accept a PR!

**Comment by CPerezz** [comment]
Hey Henry, a pleasure to work on it. 

I have some doubts since `Option` does not implement any subtle trait obviously, and I don't think we can enforce `T` to be `ConditionallySelectable` + `ConstantTimeEq`.

Therefore, I can't think of any other aproach which is not something like:
```rust
impl<T> From<CtOption<T>> for Option<T> {
    ///
    #[inline]
    fn from(source: CtOption<T>) -> Option<T> {
        if source.is_some().unwrap_u8() == 1u8 {
            Option::Some(source.value)
        } else {
            None
        }
    }
}
```

Since we're branching here, we're not enforcing ctant_time, that's for sure, but I don't know if we really care about constant time properties here. Since this will leak whether the value is `Some<T>` or `None` (if we analyze timings) as well as an `unwrap()` call will do. 

Another think I thought and that it's cleaner is the following:
```rust
impl<T> From<CtOption<T>> for Option<T>
where
    T: ConditionallySelectable,
{
    /// Requires to implement `ConditionallySelect` for `Option`
    #[inline]
    fn from(source: CtOption<T>) -> Option<T> {
        let mut tmp = CtOption::new(Option::default(), Choice::from(1u8));
        let tmp = tmp.conditional_assign(source, !source.is_some);
        
        // Always safe
        tmp.unwrap()
    }
}
```

But that would require to implement `ConditionallySelectable` for `Option`. Which Idk if it's fine or not neither if it's possible.

Do you have any other idea in mind? I tried a couple of things, but since all of the `CtOption` methods return a `CtOption<T>` it's quite difficult to do it cleaner.

**Comment by hdevalence** [comment]
@CPerezz Yes, because `Option<T>` does not have constant-time operations, it's fine to have a non-constant-time conversion from `CtOption<T>` to `Option<T>`.  So something like the first snippet would work fine.  It's OK to leak whether the value is present or not in the conversion, because the result of the conversion is a structure (`Option<T>`) that *always* leaks.

**Comment by CPerezz** [comment]
I believe this should be closed via #75
